Logical updates and tweaks to defaults

runningferret opened the pull request:
This PR attempts to bring us more in-line with Punchblocks mainline changes that have been or are expected to be merged in. Better defaults and handling of parameters.

runningferret committed:
Logical updates and tweaks to defaults

runningferret commented:
@sfgeorge @jared-prime Have a looksies. This one is relatively low priority

Jared-Prime commented on lib/punchblock/translator/asterisk/component/mrcp_recog_prompt.rb:
```diff
@@ -32,8 +32,8 @@ def validate
-            raise OptionError, "A recognition-timeout value must be a positive integer."  if @recognition_timeout && @recognition_timeout < 1
-            raise OptionError, "A speech-complete-timeout value must be a positive integer."  if @speech_complete_timeout && @speech_complete_timeout < 0
+            raise OptionError, "A recognition-timeout value must be a positive integer."  if @recognition_timeout < 1
+            raise OptionError, "A speech-complete-timeout value must be a positive integer."  if @speech_complete_timeout < 0
```
:cat:

sfgeorge replied:
Sorry, a few nit-picky things here...
- Can we change the message to "A speech-complete-timeout value must be 0 or greater."?
- We need to error upon any negative value that is not `-1`. Is this tolerable? ...
  `if @speech_complete_timeout < 0 and @speech_complete_timeout != -1`

Trying my best to meld this well with both Rayo constraints and the MRCPv2 spec

>  The value for this header field ranges from 0 to an implementation specific maximum value.

_from https://tools.ietf.org/html/draft-ietf-speechsc-mrcpv2-27#section-9.4.15_

Jared-Prime commented on lib/punchblock/translator/asterisk/component/mrcp_recog_prompt.rb:
```diff
@@ -60,8 +60,8 @@ def unimrcp_app_options
-            @recognition_timeout = input_node.recognition_timeout
-            @speech_complete_timeout = input_node.speech_complete_timeout
+            @recognition_timeout = input_node.recognition_timeout || -1
+            @speech_complete_timeout = input_node.speech_complete_timeout || 0
```
:cat2:

sfgeorge replied:
I think this will emit `sct=0` any time it is not specified, which is dangerous.  Can you add a unit test to be sure?  I think something akin to this would be good:

https://github.com/adhearsion/punchblock/blob/c26a249/spec/punchblock/translator/asterisk/component/mrcp_prompt_spec.rb#L485-L492

On a related note, [I just ping'd Ben L.](https://github.com/adhearsion/punchblock/pull/233/files#r19550952) to get a refresher on why `-1`s are used instead of nil.

runningferret commented:
@sfgeorge @Jared-Prime Don't you, da-dah, forget about meeeeee

runningferret committed:
Remove the default

Any default seems to be a bad idea.. 0 is super bad, -1 is nonsense and we don't know what the behavior will be, any non-zero positive integer is a guess at best. Best to just leave it unset.

sfgeorge commented on lib/punchblock/translator/asterisk/component/mrcp_recog_prompt.rb:
```diff
@@ -60,7 +60,7 @@ def unimrcp_app_options
             @inter_digit_timeout = input_node.inter_digit_timeout || -1
-            @recognition_timeout = input_node.recognition_timeout
+            @recognition_timeout = input_node.recognition_timeout || -1
             @speech_complete_timeout = input_node.speech_complete_timeout
```
@runningferret I hope this isn't scope-creepy, but could you default speech_complete_timeout to -1?  This will improve the ability for this to get merged upstream, for the [reasons explained by Ben L](https://github.com/adhearsion/punchblock/pull/233#discussion_r19550952).  Thanks dude!

runningferret committed:
Update defaults with PR comments

sfgeorge commented:
Even if we can't run them, I think we should have specs to go along with these changes.  Down the road when we fix the specs, we can tweak the newly added ones to work as well.  I think that this way we'll at least have something close-to-the mark, whereas otherwise we'll likely forget to write the tests after the fact.

Jared-Prime commented:
@sfgeorge @runningferret what do we want to do with this branch? should we schedule something to revisit?

sfgeorge commented:
Yeah, we can schedule something to revisit in the next sprint for sure.

> This PR attempts to bring us more in-line with Punchblocks mainline changes that have been or are expected to be merged in. Better defaults and handling of parameters.

I don't recall what Punchblock branch this referred to, and whether it was adopted.  If it wasn't or is unlikely to be adopted, I'm thinking that deviating further from Punchblock mainline isn't beneficial.  But hopefully this is the opposite, and the change was adopted.  Let me know if your memory is better than mine!

ajayanty commented:
@sfgeorge is this still required?

fyi @pseguin

runningferret commented:
:wave:

sfgeorge commented:
![hugs](https://terri0729.files.wordpress.com/2012/02/tumblr_ly8f6vo1xf1r60r74o1_400.jpg?w=820)
